lib/galaxy/model/repositories/__init__.py

```diff
@@ -0,0 +1,36 @@
+from typing import (
+    Any,
+    List,
+    Union,
+)
+
+from sqlalchemy import select
+from sqlalchemy.orm import Session
+
+from galaxy.model import Base
+from galaxy.model.store import SessionlessContext
+
+SessionType = Union[Session, SessionlessContext]
+MappedType = Base
+
+
+class BaseRepository:
```

Comment:
What is this solving ? It looks like a lot of overhead

Reply:
You mean the class `BaseRepository`? It's just a base class for all repositories which contains common functionality. So, instead of having `self.session.get(self.Foo, primary_key`) in 100+ repository classes, you have it in one.

Reply:
Can you elaborate some more on this ? I may not get it, but for instance
`parent_folder = trans.sa_session.query(trans.app.model.LibraryFolder).get(parent_id)` that you're replacing with `parent_folder = LibraryFolderRepository(trans.sa_session).get(parent_id)` does not look more efficient and I don't think is something you'd do if you're familiar with sqlalchemy (unless that's a documented new pattern ?).

It seems that this would also be problematic if you're doing further queries and mainpulations where you're expecting ORM objects.

Reply:
Sure! `session.query(Foo).get(pkey)` is SQLAlchemy-specific code. It's low level and it's tightly coupled with SA's API. When that API changes (and it did), or when we decide to use a different API (for whatever reason - like filter vs where, scalars vs. execute, etc.), we'll have to change that particular chain of method calls in every place we use it - there are many hundreds, if not thousands of those, which are also not easily grepped ("get" is so generic). With a repository, you change it in one place only. There's that.

Then there's code comprehension: whereas the `session.get(id)` vs. `foo_repository.get(id)` is not particularly convincing, any other statement is: `session.scalars(select(Foo).limit(1)).first()` to return the first Foo is less readable than `foo_repository.get_first()`. 

If one has to write low-level SA code to do any kind of data access, that makes learning low-level SA a prerequisite for writing any code that interacts with the db. Besides, being familiar with SQLAlchemy doesn't necessarily guarantee that, for example, one would know to use `limit(1)` in the case above, or remember to do that. In other words, in my opinion, wrapping library-level data access code in a repository abstraction makes the code both more maintainable (by far!) and more comprehensible.

That said, you are absolutely right about the challenge of more complex queries and manipulations: we certainly don't want to rewrite parts of SA's API! My hope is that we'll be able to tuck away the vast majority of such low level code into repositories, and eliminate a ton of duplication in the process. But yes, it's possible that there will be some logic that is too complex to be abstracted away without adding a lot of conceptual overhead - if that's the case, we'll leave such code as is, of course. I'm working my way up from trivial to simple (those are duplicated all over the code base) - to more complex, so by the time we have complex db logic to untangle, there'll be a ton of improvement done already.

Data access repositories is a pattern. I'll provide a more detailed description as it relates to our code base in the PR description. Essentially, it boils down to eliminating duplication and encapsulating low level code/logic in objects and methods with names that describe *what* is being done/returned (e.g. foo repo returning all deleted bars) as opposed to *how* it is implemented (e.g. session/scalars/select/first/etc...)

Reply:
> tuck away the vast majority of such low level code into repositories

that IMO is not the right direction. I am -0.9 on this change, sorry.

Reply:
I'm... persuadable in either direction on introducing this pattern... but I will say the tool parameter sql alchemy usage is a good example of why the pattern might be useful. I like the change of moving low-level logic out of there and I don't think managers would be a great fit because I think they belong above that level of the dependency chain.

My concern with adding the abstraction layer would be... are we making simple things a little more complex and not helping complex things at all? I think a lot of the base manager stuff is problematic in this way - a deep layer of complexity that only works for 3 object types and just makes the rest of the code feel inconsistent.

Reply:
I'm sorry, I don't think you will be able to write something to convince me while you have code like this:
```diff
class QuotaRepository(BaseRepository):
    def __init__(self, session: SessionType):
        super().__init__(session, Quota)

    def get(self, primary_key: int) -> Quota:
        return cast(Quota, super().get(primary_key))
```

Anything is better than that. If this is about providing `get_first` and other helpers in a convenient way you could subclass or compose a different session type.

Reply:
I know it's still draft and so not entirely fair to draw conclusions, but

> def get(self, primary_key: int) -> Quota:

looks wrong to me, since get returns `Model | None`. That leads me to worry about insulating ourselves from proper types

Reply:
> > def get(self, primary_key: int) -> Quota:
> 
> looks wrong to me, since get returns `Model | None`. That leads me to worry about insulating ourselves from proper types

That's not the intent, of course. The correct return type should be `Optional[Quota]` (so here's one more iteration).

Reply:
> That gives you much tighter type checking

to me these are highly questionable redundant annotations, given that everything returned through the session already has the proper type:

```
# (variable) users1: Sequence[User]
users1 = session.scalars(select(User)).all()

# (variable) user: User
user = session.query(User).one()

# (variable) user_iter: Iterator[User]
user_iter = iter(session.scalars(select(User)))
```
see https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#sql-expression-typing-examples



> session.execute(select(Foo)).scalar_one()	foo_repo.one()

that's not always going to be what we want for something called `.one()`, so why not roll with the upstream syntax ? It's not like the sqlalchemy people hate good syntax / UX.

Reply:
> Here are just a few examples:

These are not complicated examples IMO - these are exactly the simple code.

Reply:
Something like ``get_with_filter_order_by_hid`` - is indeed code that I think belongs in the model layer and should be tested in isolation. I think though just a static method on the HDA class or a function in that module that took in the sa_session along with the rest of the arguments you've created and just did the work directly would be simpler and feel more "python" to me.

Reply:
Yes, you're right - those are quite simple. I'll move on to more complex ones - I hope I'll find better examples.

The `get_with_filter_order_by_hid` - bad approach altogether: I'm doing exactly what we shouldn't do - taking perfectly normal SA syntax and restating the same imperative logic as an ugly method name which doesn't improve anything beyond reducing duplicate code. Just an iteration - so please ignore. 

I think I have a better option - just pushed. The method name, I think, should say *what* is being fetched/done. The repo module (or some module within the model) contains all the low level SA code that details *how* it's done.  See [`repository.hda`](https://github.com/galaxyproject/galaxy/pull/16434/files#diff-fbf8c38da6c2f286bd70d2fb60c2178981f10b765ad66566d0baa247bb53bc16R15) and [`webapps.galaxy.services.history.py`](https://github.com/galaxyproject/galaxy/pull/16434/files#diff-13b6a7251d924008c6cfe1de85cbabc93dc77f12f32c98a059914843b82fcd64R644). Naturally, there will be a lot of similar methods with different names - that can be then refactored in the repositories to eliminate any duplicate constructs.